docs/base/environment_variables.rst
=====================
Environment Variables
=====================

Below is a reference list of environment variables used to configure portl. Many of these are described in the Installation section. Some are described within the document in relevant sections. Configurations in the Administration Panel also exist as environment variables for first time startup; if these are desired changed after initial startup and database creation, they must be changed via the Administration Panel UI.

Complete List of Environment Variables
--------------------------------------

::

  - export portl_CONFIG_MODE=Production
  - export portl_SECRET_KEY=your_secret_key123
  - export USE_VAULT=1
  - export VAULT_ADDR=http://127.0.0.1:8200
  - export VAULT_TOKEN=e1c70d27-7c7f-6f6a-fb18-b0c0382667b7
  - export UNSEAL_VAULT=1
  - export UNSEAL_VAULT_KEY1=nvn7aJgA1wA/z2en/rqeNxU8zxSNcl8bH6L4Voch7LQQ
  - export UNSEAL_VAULT_KEY2=Xndum7gpEykrsRAf6UkaDOSdGFqNMuswggD7zdIYakwI
  - export UNSEAL_VAULT_KEY3=+17lQib+Z/MP5I30Fhd9/yoox9XKzk8bWERv9v3nZ5Ow
  - export portl_SERVER_ADDR=https://10.10.10.5
  - export GOTTY_PORT_REDIRECTION=1
  - export GOTTY_BYPASS_KEY_PROMPT=1
  - export GOTTY_START_PORT=9000
  - export GOTTY_END_PORT=9100
  # Uncomment for Postgres Mode:
  # - export portl_DATABASE_URL=postgresql://portl:password@localhost:5432/portl
  # - export POSTGRES_PASSWORD=password
  # - export POSTGRES_USER=portl
  # - export POSTGRES_HOST=localhost
  # - export POSTGRES_PORT=5432
  # - export POSTGRES_DB=portl
  # Uncomment for MySQL Mode:
  # - export portl_DATABASE_URL=mysql://root:password@localhost/portl
  # Uncomment for SQLite Mode:
  - export portl_DATABASE_URL=sqlite:///database.db?check_same_thread=False
  - export SQLITE_WAL_MODE=1     (to enable multiprocessing support in sqlite)
  - export PATH_CUSTOM_PROPERTIES=/home/user/portl/portl_properties.yml
  - export CREATE_EXAMPLES=0
  - export MAIL_SERVER=smtp.company.com
  - export MAIL_PORT=25
  - export MAIL_USE_TLS=0
  - export MAIL_USERNAME=portl-user
  - export MAIL_PASSWORD=portl-user
  - export MAIL_SENDER=portl@company.com
  - export MAIL_RECIPIENTS=portl-user@company.com
  - export MATTERMOST_URL=https://mattermost.company.com/hooks/i1phfh6fxjfwpy586bwqq5sk8w
  - export MATTERMOST_CHANNEL=
  - export MATTERMOST_VERIFY_CERTIFICATE=0
  - export USE_LDAP=1
  - export LDAP_SERVER=ldap://domain.ad.company.com
  - export LDAP_USERDN=domain.ad.company.com
  - export LDAP_BASEDN=DC=domain,DC=ad,DC=company,DC=com
  - export LDAP_ADMIN_GROUP=portl.Users,network.Admins
  - export USE_TACACS=1
  - export TACACS_ADDR=122.10.10.10
  - export TACACS_PASSWORD=tacacs_password
  - export USE_SYSLOG=0
  - export SYSLOG_ADDR=0.0.0.0
  - export SYSLOG_PORT=514
  - export CLUSTER=0
  - export CLUSTER_ID=1
  - export CLUSTER_SCAN_SUBNET=192.168.0.0
  - export CLUSTER_SCAN_PROTOCOL=https
  - export CLUSTER_SCAN_TIMEOUT=0.05
  - export DEFAULT_LONGITUDE=-96.0
  - export DEFAULT_LATITUDE=33.0
  - export DEFAULT_ZOOM_LEVEL=5
  - export DEFAULT_VIEW=2D  (or 2DC or 3D)
  - export DEFAULT_MARKER=Circle  (or Image or "Circle Marker")
  - export GIT_AUTOMATION=git@gitlab.company.com:portl-group/portl-automation-data.git
  - export GIT_CONFIGURATIONS=git@gitlab.company.com:portl-group/portl-configurations-data.git
  - export POOL_FILTER=All objects
  - export portl_LOG_LEVEL=DEBUG
  - export GUNICORN_ACCESS_LOG=logs/app_logs/access.log
  - export GUNICORN_LOG_LEVEL=debug

docs/inventory/configuration_management.rst
========================
Configuration Management
========================

portl can work as a network device configuration backup tool and replace Oxidized/Rancid with the following features:
  - Poll network elements; download configurations when they change
  - Easily view the current configuration of a device in the inventory
  - Search for any text in any configuration
  - View differences between various revisions of a configuration
  - Download device configuration to a local text file
  - Use the ReST API support to return a specified device's configuration
  - Export all device configurations to a remote Git repository (e.g. Gitlab)

Device configuration
--------------------

All devices are listed in the :guilabel:`inventory/configuration_management` page. By default, configurations are retrieved by a service called ``Configuration Backup Service``, which:
  - Uses Netmiko to fetch the configuration
  - Updates the device ``configuration`` property (a python dictionary that contains the most recent configurations)
  - Writes the configuration to a local text file (located in portl/git/configurations)

.. image:: /_static/inventory/configuration_management/device_configuration.png
   :alt: Configuration Management table.
   :align: center

For some devices, the configuration cannot be retrieved with only a netmiko command. You can create your own configuration backup service(s) if need be. Targets are defined at the service level, like any other services.
The task ``poller_task`` will periodically run all services whose ``configuration_backup_service`` parameter is set to ``True``, as shown in the default configuration backup service here: https://github.com/afourmy/portl/blob/master/portl/automation/services/configuration_management/configuration_backup_service.py#L15

Configure polling
-----------------

The polling process is controlled by the ``Poller`` task. The ``Poller`` task is configured to run the ``Configuration Management Workflow``.

.. image:: /_static/inventory/configuration_management/configuration_management_workflow.png
   :alt: Configuration Management Workflow.
   :align: center

To start the configuration polling mechanism, you need to go to the :guilabel:`scheduling/task_management` page and start the ``Poller`` task by pressing the ``Resume`` button.
By default, the ``Poller`` task will run every hour (3600 seconds), but you can change the frequency from the ``Edit`` form.

Search and display the configuration
------------------------------------

You can search for a specific word in the current configuration of all devices with the input field on top of the ``Configuration`` column. portl will filter the list of devices based on whether the current configuration of the device contains this word.

By clicking on the ``Configuration`` button, you can display and compare the device configurations.

.. image:: /_static/inventory/configuration_management/display_configuration.png
   :alt: Display Configuration.
   :align: center

All runs are stored in the ``Display`` and ``Compare With`` pull-down lists:
  - Selecting a run from ``Display`` will display the associated configuration.
  - Selecting a run from ``Compare With`` will compare the configuration with the one selected in ``Display``.

Additionally, you can click on ``Raw logs`` to open a pop up that contains only the configuration (useful for copy/pasting), and click on ``Clear`` to remove all previously stored configurations from the database.

Comparing two configurations will display a git-like line-by-line diff similar to the one below:

.. image:: /_static/inventory/configuration_management/compare_configurations.png
   :alt: Compare Configurations.
   :align: center

Advanced
--------

- Number of configurations stored in the database: by default, portl stores the 10 most recent configurations in the database. The polling process is controlled by the ``configuration_backup`` service. You can change the number of stored configurations by changing the ``Number of configurations stored`` property.
- Configurations are retrieved with Netmiko. By default, portl uses the driver defined at device level to run the command. You can use a driver configured at service level instead, by unticking the ``Use driver from device`` check box.
- Configurations are pushed to the git 'configurations' repository automatically. This is not to be confused with the 'Push to git' option for Services and Workflows, which enables pushing the Job logs into the 'automation' repository

portl/config.py
from os import environ
from typing import Dict, Type


class Config(object):
    # SQL Alchemy
    SQLALCHEMY_ENGINE_OPTIONS = {"pool_pre_ping": True}
    SQLALCHEMY_RECORD_QUERIES = True
    SQLALCHEMY_TRACK_MODIFICATIONS = False

    # portl
    PORTL_SERVER_ADDR = environ.get("PORTL_SERVER_ADDR")

    # WebSSH (GoTTY)
    GOTTY_PORT_REDIRECTION = int(environ.get("GOTTY_PORT_REDIRECTION", False))
    GOTTY_BYPASS_KEY_PROMPT = environ.get("GOTTY_BYPASS_KEY_PROMPT")
    GOTTY_START_PORT = int(environ.get("GOTTY_START_PORT", 9000))
    GOTTY_END_PORT = int(environ.get("GOTTY_END_PORT", 9100))

    # Vault
    USE_VAULT = int(environ.get("USE_VAULT", False))

    # LDAP
    LDAP_SERVER = environ.get("LDAP_SERVER")
    LDAP_USERDN = environ.get("LDAP_USERDN")
    LDAP_BASEDN = environ.get("LDAP_BASEDN")
    LDAP_ADMIN_GROUP = environ.get("LDAP_ADMIN_GROUP", "").split(",")

    # TACACS+
    USE_TACACS = int(environ.get("USE_TACACS", False))
    TACACS_ADDR = environ.get("TACACS_ADDR")
    TACACS_PASSWORD = environ.get("TACACS_PASSWORD")

    # Syslog
    USE_SYSLOG = int(environ.get("USE_SYSLOG", False))
    SYSLOG_ADDR = environ.get("SYSLOG_ADDR", "0.0.0.0")
    SYSLOG_PORT = int(environ.get("SYSLOG_PORT", 514))

    # Examples
    CREATE_EXAMPLES = int(environ.get("CREATE_EXAMPLES", False))

    # Custom Services
    CUSTOM_SERVICES_PATH = environ.get("CUSTOM_SERVICES_PATH")

    # Notifications
    # - via Mail
    MAIL_SERVER = environ.get("MAIL_SERVER", "smtp.googlemail.com")
    MAIL_PORT = int(environ.get("MAIL_PORT", "587"))
    MAIL_USE_TLS = int(environ.get("MAIL_USE_TLS", True))
    MAIL_USERNAME = environ.get("MAIL_USERNAME")
    MAIL_PASSWORD = environ.get("MAIL_PASSWORD")
    MAIL_SENDER = environ.get("MAIL_SENDER", "portl@portl.fr")
    MAIL_RECIPIENTS = environ.get("MAIL_RECIPIENTS", "")

    # - via Mattermost
    MATTERMOST_URL = environ.get("MATTERMOST_URL", "")
    MATTERMOST_CHANNEL = environ.get("MATTERMOST_CHANNEL", "")
    MATTERMOST_VERIFY_CERTIFICATE = int(
        environ.get("MATTERMOST_VERIFY_CERTIFICATE", True)
    )

    # - Via Slack
    SLACK_TOKEN = environ.get("SLACK_TOKEN", "")
    SLACK_CHANNEL = environ.get("SLACK_CHANNEL", "")

    # Logging
    PORTL_LOG_LEVEL = environ.get("PORTL_LOG_LEVEL", "DEBUG").upper()

    # Cluster
    # In production, for scalability and high-availability purposes, it is
    # recommended to deploy not one, but multiple instances of portl.
    CLUSTER = int(environ.get("CLUSTER", False))
    CLUSTER_ID = int(environ.get("CLUSTER_ID", True))
    CLUSTER_SCAN_SUBNET = environ.get("CLUSTER_SCAN_SUBNET", "192.168.105.0/24")
    CLUSTER_SCAN_PROTOCOL = environ.get("CLUSTER_SCAN_PROTOCOL", "http")
    CLUSTER_SCAN_TIMEOUT = float(environ.get("CLUSTER_SCAN_TIMEOUT", 0.05))

    # Geographical Parameters
    DEFAULT_LONGITUDE = float(environ.get("DEFAULT_LONGITUDE", -96.0))
    DEFAULT_LATITUDE = float(environ.get("DEFAULT_LATITUDE", 33.0))
    DEFAULT_ZOOM_LEVEL = int(environ.get("DEFAULT_ZOOM_LEVEL", 5))
    DEFAULT_VIEW = environ.get("DEFAULT_VIEW", "2D")
    DEFAULT_MARKER = environ.get("DEFAULT_MARKER", "Image")

    # Git Parameters
    GIT_AUTOMATION = environ.get("GIT_AUTOMATION", "")
    GIT_CONFIGURATIONS = environ.get("GIT_CONFIGURATIONS", "")

    # Pool Filter Parameter
    POOL_FILTER = environ.get("POOL_FILTER", "All objects")


class DebugConfig(Config):
    DEBUG = True
    SECRET_KEY = environ.get("PORTL_SECRET_KEY", "get-a-real-key")

    # Database
    SQLALCHEMY_DATABASE_URI = environ.get(
        "PORTL_DATABASE_URL", "sqlite:///database.db?check_same_thread=False"
    )

    # Mail
    MAIL_DEBUG = 1


class ProductionConfig(Config):
    DEBUG = False
    # In production, the secret MUST be provided as an environment variable.
    SECRET_KEY = environ.get("PORTL_SECRET_KEY")

    # Database
    SQLALCHEMY_DATABASE_URI = environ.get(
        "PORTL_DATABASE_URL",
        "postgresql://{}:{}@{}:{}/{}".format(
            environ.get("POSTGRES_USER", "portl"),
            environ.get("POSTGRES_PASSWORD", "portl"),
            environ.get("POSTGRES_HOST", "db"),
            environ.get("POSTGRES_PORT", 5432),
            environ.get("POSTGRES_DB", "portl"),
        ),
    )

    # Vault
    # In production, all credentials (hashes, usernames and passwords) are
    # stored in a vault.
    # There MUST be a Vault to use portl in production mode securely.
    USE_VAULT = int(environ.get("USE_VAULT", True))
    VAULT_ADDR = environ.get("VAULT_ADDR")
    VAULT_TOKEN = environ.get("VAULT_TOKEN")
    UNSEAL_VAULT = environ.get("UNSEAL_VAULT")
    UNSEAL_VAULT_KEY1 = environ.get("UNSEAL_VAULT_KEY1")
    UNSEAL_VAULT_KEY2 = environ.get("UNSEAL_VAULT_KEY2")
    UNSEAL_VAULT_KEY3 = environ.get("UNSEAL_VAULT_KEY3")
    UNSEAL_VAULT_KEY4 = environ.get("UNSEAL_VAULT_KEY4")
    UNSEAL_VAULT_KEY5 = environ.get("UNSEAL_VAULT_KEY5")

    # Security
    SESSION_COOKIE_HTTPONLY = True
    REMEMBER_COOKIE_HTTPONLY = True
    REMEMBER_COOKIE_DURATION = 3600


config_dict: Dict[str, Type[Config]] = {
    "Production": ProductionConfig,
    "Debug": DebugConfig,
}
